lib/icingadb/redisconnection.cpp

```diff
@@ -338,21 +338,18 @@ void RedisConnection::WriteLoop(asio::yield_context& yc)
 	for (;;) {
 		m_QueuedWrites.Wait(yc);
 
-	WriteFirstOfHighestPrio:
-		for (auto& queue : m_Queues.Writes) {
-			if (m_SuppressedQueryKinds.find(queue.first) != m_SuppressedQueryKinds.end() || queue.second.empty()) {
-				continue;
-			}
-
-			auto next (std::move(queue.second.front()));
-			queue.second.pop();
+		while (!m_Queues.Writes.empty() && m_SuppressedQueryKinds.find(m_Queues.Writes.top().Priority) == m_SuppressedQueryKinds.end()) {
```

Comment:
This now changes the behavior how suppression works. Previously it would continue with lower priorities but now it will just stop if the highest queued priority is suppressed.

Currently, only the priorities `CheckResult` and `State` are suppressed and always at the same time:
https://github.com/Icinga/icinga2/blob/171a753d9b0f7284ca77255ef64df91431dd2832/lib/icingadb/icingadb.cpp#L84-L85
https://github.com/Icinga/icinga2/blob/171a753d9b0f7284ca77255ef64df91431dd2832/lib/icingadb/icingadb-objects.cpp#L138-L144

However these are not the two lowest priorities, there's `History` in between:
https://github.com/Icinga/icinga2/blob/171a753d9b0f7284ca77255ef64df91431dd2832/lib/icingadb/redisconnection.hpp#L60-L67

So if there is a `State` write in the queue, `History` writes are no longer executed even though they're suppressed.

We should decide if this behavior is acceptable and if it is, change the suppression mechanism accordingly, i.e. make it a single priority and everything with this or lower priority is suppressed.

Reply:
@Yonas-net Figure out whether/how this is possible.

Reply:
If you add to the above [example](https://github.com/Icinga/icinga2/pull/8557#issuecomment-759412404) the following example then it happens exactly what @julianbrost explained.

```cxx
std::set<QueryPriority> m_SuppressedQueryKinds;
m_SuppressedQueryKinds.emplace(QueryPriority::CheckResult);
m_SuppressedQueryKinds.emplace(QueryPriority::State);

while (!Writes.empty() && m_SuppressedQueryKinds.find(Writes.top().Priority) == m_SuppressedQueryKinds.end()) {
	cout << Writes.top().Key << "  " << Writes.top().order << endl;
	Writes.pop();
}
```

**Output**:
```cxx
Heartbeat1  5
Heartbeat2  6
Heartbeat3  9
Heartbeat4 12
Config1  0
Config2  7
```

History would normally also have to be output but that is no longer there now.